fahrschein/src/main/java/org/zalando/fahrschein/NakadiClient.java

```diff
@@ -110,17 +110,23 @@ public <T> void publish(String eventName, List<T> events) throws EventPublishing
 
         Response response = null;
         try {
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.onPublish(eventName, events));
+            eventPublishingHandlers.forEach(handler -> handler.onPublish(eventName, events));
             response = request.execute();
             LOG.debug("Successfully published [{}] events for [{}]", events.size(), eventName);
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.afterPublish());
         } catch (Throwable t) {
-            eventPublishingHandlers.forEach(requestHandler -> requestHandler.onError(events, t));
+            eventPublishingHandlers.descendingIterator().forEachRemaining(handler -> handler.onError(events, t));
```

Comment:
If the on-error method creates an exception, we would lose the original error, so at least we should log something before executing the handlers.

Reply:
Maybe we should even try/catch the on error execution and just log something if the handler execution failed.